### PR TITLE
Login Rework: new login flow for WPCOM re-auth

### PR DESCRIPTION
**Pull request by hypest**
**Note: #6164 needs to be merged first.**

This PR enables the new login flow for the WordPress.com re-auth entry point. This entry point is triggered when the app detects that the auth token is no longer valid, at which point it prompts the user to log in again.

Since the prompt is for asking the user to log into WPCOM, the selfhosted flows are not offered. At the epilogue screen, only the "Continue" button is offered.

The final visual touches, the messages and final texts will be done in a future PR.

To test:
* Have the app logged in to WordPress.com and just put it in the background
* Use Calypso on the web and go to https://wordpress.com/me/security/connected-applications. From there, disconnect the Android app from the account.
* Return to the mobile app and relaunch it
* App should ask to login again. Continue and log in using the same WPCOM account.
* Epilogue screen should appear on successful login

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/NewUserFragment.java [hunk 1]
```diff
@@ -464,6 +487,26 @@ public void onClick(View v) {
...
+            } else {
+                throw new RuntimeException(context.toString() + " must implement LoginListener");
+            }
+        }
+    }
+
+    @Override
+    public void onDetach() {
+        super.onDetach();
+
```
Comment (theck13) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/NewUserFragment.java]: This newline is unnecessary.

File: WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java [hunk 2]
```diff
@@ -589,9 +589,7 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
                 }
                 break;
             case RequestCodes.REAUTHENTICATE:
-                if (resultCode == RESULT_CANCELED) {
```
Comment (theck13) [inline WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java]: Is the `RESULT_OK` case the only possibility?

**Review by theck13 — COMMENTED** [review]
While testing the following step, the "Notifications couldn't be refreshed as this time" toast message was displayed as shown in the screenshot below.  That seemed a little confusing.

> * Return to the mobile app and relaunch it

6181_toast

I also had one question about the `WPMainActivity` result code possibilities.  Other than that, I had only a single minor nitpick.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/NewUserFragment.java (hunk 1 above)
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/NewUserFragment.java]: Handled in a different PR (https://github.com/wordpress-mobile/WordPress-Android/pull/6164#discussion_r126885786) so, I'll leave as is in this one.

File: WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java (hunk 2 above)
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java]: Yes. `RESULT_OK` happens if login completed, in which case this Activity just continues and starts the GCMRegistrationIntentService service.

If Login gets cancelled, app just returns to the main activity. Note, with the current implementation the login screen will be relaunched anyway, unless there's no connectivity.

**Comment by hypest** [comment]
Hey @theck13 , 

> ... the "Notifications couldn't be refreshed as this time" toast message was displayed as shown in the screenshot below. That seemed a little confusing.

Yes, I'm afraid that's a bit of an annoyance in this case. It's present in the production app as well and is not caused by this PR so, I'd recommend we leave is out of this PR's scope. WDYT?

Beyond that, I've addressed all the other comments without any code changes so, let me know if there's anything more I can do for this PR. Thanks!

**Comment by theck13** [comment]
> Yes, I'm afraid that's a bit of an annoyance in this case. It's present in the production app as well and is not caused by this PR so, I'd recommend we leave is out of this PR's scope. WDYT?

In that case, leaving it out makes sense with me.

It looks like the merge of https://github.com/wordpress-mobile/WordPress-Android/pull/6164 caused a merge conflict.  Once that's resolved, it will be ready to ship.

**Comment by hypest** [comment]
Conflict resolved @theck13 , ready for another pass, thanks!

**Review by theck13 — APPROVED** [review]
:shipit:

**Pull request merged**